cpp/src/fil/fil.cu

```diff
@@ -727,71 +867,88 @@ void node2fil_dense(std::vector<dense_node>* pnodes,
-    dense_node({}, val_t{.f = threshold}, tree.SplitIndex(node_id), default_left, false, false);
   int left = 2 * cur + 1;
-  node2fil_dense(pnodes, root, left, tree, tl_left, forest_params, vector_leaf, leaf_counter);
-  node2fil_dense(pnodes, root, left + 1, tree, tl_right, forest_params, vector_leaf, leaf_counter);
+  conversion_state<dense_node> cs =
+    tl2fil_inner_node<dense_node>(left, tree, node_id, forest_params, cat_sets, bit_pool_offset);
+  (*pnodes)[root + cur] = cs.node;
+  node2fil_dense(pnodes,
+                 root,
+                 left,
+                 tree,
+                 cs.tl_left,
+                 forest_params,
+                 vector_leaf,
+                 leaf_counter,
+                 cat_sets,
+                 bit_pool_offset);
+  node2fil_dense(pnodes,
+                 root,
+                 left + 1,
+                 tree,
+                 cs.tl_right,
+                 forest_params,
+                 vector_leaf,
+                 leaf_counter,
+                 cat_sets,
+                 bit_pool_offset);
 }
 
 template <typename T, typename L>
 void tree2fil_dense(std::vector<dense_node>* pnodes,
                     int root,
                     const tl::Tree<T, L>& tree,
+                    std::size_t tree_idx,
                     const forest_params_t& forest_params,
                     std::vector<float>* vector_leaf,
-                    size_t* leaf_counter)
+                    size_t* leaf_counter,
+                    cat_sets_owner* cat_sets)
 {
-  node2fil_dense(pnodes, root, 0, tree, tree_root(tree), forest_params, vector_leaf, leaf_counter);
+  node2fil_dense(pnodes,
+                 root,
+                 0,
+                 tree,
+                 tree_root(tree),
+                 forest_params,
+                 vector_leaf,
+                 leaf_counter,
+                 cat_sets,
+                 &cat_sets->bit_pool_offsets[tree_idx]);
 }
 
 template <typename fil_node_t, typename T, typename L>
-int tree2fil_sparse(std::vector<fil_node_t>& nodes,
-                    int root,
-                    const tl::Tree<T, L>& tree,
-                    const forest_params_t& forest_params,
-                    std::vector<float>* vector_leaf,
-                    size_t* leaf_counter)
+__noinline__ int tree2fil_sparse(std::vector<fil_node_t>& nodes,
+                                 int root,
+                                 const tl::Tree<T, L>& tree,
+                                 std::size_t tree_idx,
+                                 const forest_params_t& forest_params,
+                                 std::vector<float>* vector_leaf,
+                                 size_t* leaf_counter,
+                                 cat_sets_owner* cat_sets)
 {
   typedef std::pair<int, int> pair_t;
   std::stack<pair_t> stack;
   int built_index = root + 1;
   stack.push(pair_t(tree_root(tree), 0));
+  size_t bit_pool_offset = cat_sets->bit_pool_offsets[tree_idx];
```

Comment:
```suggestion
  std::size_t bit_pool_offset = cat_sets->bit_pool_offsets[tree_idx];
```

Reply:
Nope, this is primarily about what it communicates to future developers. `std::size_t` is a C++ type; `size_t` is a C type. By sticking to C++ types generally but using C types where we expect to interface with C code, it becomes clearer where we expect those seams to be. Also, the C++ standard says that `size_t` may be removed from later versions of the standard, though I would be shocked if that actually happens.